### PR TITLE
Ratelimiter

### PR DESCRIPTION
**Pull request by kyriakid1s**
### Description:
Implement a Token Bucket rate limiter that supports:

-Non-blocking token checks with Allow() (returns whether operation can proceed immediately).

-Blocking wait with Wait().

-Adjustable capacity (burst size) and refill rate (tokens per second).

-Optionally, support other algorithms like Fixed Window or Leaky Bucket for flexibility.

### Checklist:  
- [x] **Tests Passing**: Verify by running `make test`.  
- [x] **Golint Passing**: Confirm by running `make lint`.  

#### If added a new utility function or updated existing function logic:
* [ ] Updated the utility package `EXAMPLES.md`
* [ ] Updated the utility package `README.md`

#### If added a new utility package:
* [x] Updated the main `README.md` utility packages table

## Summary by CodeRabbit

- **New Features**
  - Added a thread-safe token-bucket rate limiter with burst support, non-blocking checks, blocking waits that honor context cancellation, runtime capacity/refill adjustments, and token introspection (no public API changes).

- **Documentation**
  - Added package overview and a comprehensive examples guide demonstrating usage patterns, edge cases, blocking behavior, timeouts, and high-concurrency scenarios; updated top-level README links.

- **Tests**
  - Added extensive tests covering concurrency, timing/refill behavior, config changes, edge cases, cancellation, and robustness.

- **Chores**
  - Updated .gitignore to ignore .DS_Store.

### CODE REVIEW

**Review by kashifkhan0771 — APPROVED** [review]
The changes look good to me. Once you resolve the linter issue, we should be good to merge this feature

**Comment by kashifkhan0771** [comment]
@coderabbitai review

File: ratelimiter/ratelimiter.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+func (t *TokenBucket) WaitN(ctx context.Context, n int) error {
+	if n <= 0 {
+		return nil
+	}
+
+	for {
+		now := time.Now()
+		t.mu.Lock()
+		if n > t.capacity {
+			currencyCap := t.capacity
```
Comment (shahzadhaider1) [inline ratelimiter/ratelimiter.go:93]: This looks like a typo (currencyCap -> currentCap), or is it intentional?

File: ratelimiter/ratelimiter.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+// SetCapacity adjusts the bucket capacity at runtime.
+// If the new capacity is smaller than the current number of tokens,
+// the token count is reduced to match the new capacity.
+// This method is thread-safe.
+// SetCapacity adjusts the bucket capacity at runtime.
+// If the new capacity is smaller than the current number of tokens,
+// the token count is reduced to match the new capacity.
+// This method is thread-safe.
```
[the remark this answers is not in the document]
Comment (kyriakid1s) [reply ratelimiter/ratelimiter.go:162]: Yes you are right. Fixed

File: ratelimiter/ratelimiter.go (hunk 1 above)
Comment (kyriakid1s) [reply ratelimiter/ratelimiter.go:93]: This is typo. Fixed

File: ratelimiter/helper.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	if n > t.capacity || t.refillRate < 0 {
+		return -1
+	}
+
+	if t.tokens >= float64(n) {
+		return 0
+	}
+
+	needed := float64(n) - t.tokens
+	secs := needed / t.refillRate
```
[the remark this answers is not in the document]
Comment (kyriakid1s) [reply ratelimiter/helper.go:49]: If the user sets refillRate to less than or equal to 0, it defaults to 1.

I can add an if statement to check it before performing the division.

**Review by kashifkhan0771 — APPROVED** [review]
